toucan_connectors/wootric/wootric_connector.py

```diff
@@ -0,0 +1,141 @@
+import asyncio
+from datetime import datetime, timedelta
+from itertools import chain
+import json
+from typing import List, Optional
+
+from aiohttp import ClientSession
+import requests
+
+import pandas as pd
+
+from toucan_connectors.toucan_connector import ToucanConnector, ToucanDataSource
+
+_TOKEN_CACHE = None  # internal cache to avoid re-requesting OAUTH access_token
+
+
+async def fetch(session, url):
+    """aiohttp version of requests.get(...).json()"""
+    async with session.get(url) as response:
+        return json.loads(await response.read())
+
+
+async def _batch_fetch(urls):
+    """fetch asyncrhonously `urls` in a single batch"""
+    async with ClientSession() as session:
+        tasks = (asyncio.Task(fetch(session, url)) for url in urls)
+        return await asyncio.gather(*tasks)
+
+
+def batch_fetch(urls):
+    """fetch asyncrhonously `urls` in a single batch"""
+    loop = asyncio.get_event_loop()
+    future = asyncio.ensure_future(_batch_fetch(urls))
+    return loop.run_until_complete(future)
+
+
+def fetch_wootric_data(query, props_fetched=None, batch_size=5, max_pages=30):
+    """call the `query` wootric API endpoint and handle pagination
+
+    Parameters:
+
+    - `query`: the API endpoint, e.g. `'response'`
+
+    - `props_fetched`: if specified, a list of properties to pick in the json documents
+      returned by wootric
+
+    - `batch_size`: number of requests to batch together. i.e. if `batch_size=5`, then the
+      API will be queried by batches of 5 queries until data is exhausted or `max_pages`
+      is reached.
+
+    - `max_pages`: maximum number of pages to crawl.
+    """
+    all_data = []
+    for page in range(1, max_pages + 1, batch_size):
+        lastpage = min(page + batch_size, max_pages + 1)
+        urls = [f'{query}&page={pagenum}' for pagenum in range(page, lastpage)]
+        responses = batch_fetch(urls)
+        data = chain.from_iterable(responses)
+        if props_fetched is None:
+            all_data.extend(data)
+        else:
+            all_data.extend([
+                {prop: d[prop] for prop in props_fetched}
+                for d in data
+            ])
+        # last response is empty, it means that wootric doesn't have any data left
+        if not responses[-1]:
+            break
+    return all_data
+
+
+def access_token(connector):
+    """return OAUTH access token for connector `connector`
+
+    This function handles a cache internally to avoid re-requesting the token
+    if the one is cached is still valid.
+    """
+    global _TOKEN_CACHE
+    if _TOKEN_CACHE is not None:
+        token_infos = _TOKEN_CACHE[0]
```

Comment:
Why is it a list btw?

Reply:
A remainder of a previous implementation